### ISSUE TITLE
二郎神Pretrain的细节

### ISSUE THREAD

**Issue by VERSPD0** [IDEA-CCNL/Fengshenbang-LM#39]
请问是否有二郎神预训练的实验细节呢，谢谢！

**Comment by ganzhiruyi** [comment]
暂时还没有开放二郎神pretrain的细节，后续我们也会开放各个模型的训练脚本和参数，目前参数可以参考：
`
--num-layers 24 \
--hidden-size 2048 \
--num-attention-heads 8 \
--micro-batch-size 4 \
--global-batch-size 1024 \
--seq-length 512 \
--max-position-embeddings 512 \
--train-iters 1800000 \
--lr 0.0001 \
--lr-decay-style linear \
--min-lr 1.0e-5 \
--lr-decay-iters 1600000 \
--weight-decay 1e-2 \
--clip-grad 1.0 \
--lr-warmup-fraction .01 \
--eval-iters 20 \
--tokenizer-type BertCNWWMTokenizer \
--fp16 "
`

**Comment by yayaQAQ** [comment]
@ganzhiruyi 
请问怎么将  32张A100  连接起来训练的呢？我们想在我们的语料上继续预训练这个模型，但是不太了解怎么这么多卡联合训练。

**Comment by ganzhiruyi** [comment]
> @ganzhiruyi 请问怎么将 32张A100 连接起来训练的呢？我们想在我们的语料上继续预训练这个模型，但是不太了解怎么这么多卡联合训练。

可以尝试下我们的封神框架，支持多卡训练，参考这个https://github.com/IDEA-CCNL/Fengshenbang-LM/blob/main/fengshen/examples/pretrain_erlangshen/pretrain_erlangshen_1.3B.sh

**Comment by bigprince97** [comment]
请问下，训练了1800000步是将数据重复了几个epoch，使用了什么优化器？

**Comment by yayaQAQ** [comment]
> > @ganzhiruyi 请问怎么将 32张A100 连接起来训练的呢？我们想在我们的语料上继续预训练这个模型，但是不太了解怎么这么多卡联合训练。
> 
> 可以尝试下我们的封神框架，支持多卡训练，参考这个https://github.com/IDEA-CCNL/Fengshenbang-LM/blob/main/fengshen/examples/pretrain_erlangshen/pretrain_erlangshen_1.3B.sh

我尝试了一下，我不太懂为什么会需要一个npy文件，是如何构建这个npy文件的呢？
<img width="1338" alt="image" src="https://user-images.githubusercontent.com/20497531/171527676-14a84fb5-19c4-479c-8a83-5fabb876a5d6.png">

**Comment by GGGGGGXY** [comment]
> > > @ganzhiruyi 请问怎么将 32张A100 连接起来训练的呢？我们想在我们的语料上继续预训练这个模型，但是不太了解怎么这么多卡联合训练。
> > 
> > 
> > 可以尝试下我们的封神框架，支持多卡训练，参考这个https://github.com/IDEA-CCNL/Fengshenbang-LM/blob/main/fengshen/examples/pretrain_erlangshen/pretrain_erlangshen_1.3B.sh
> 
> 我尝试了一下，我不太懂为什么会需要一个npy文件，是如何构建这个npy文件的呢？ <img alt="image" width="1338" src="https://user-images.githubusercontent.com/20497531/171527676-14a84fb5-19c4-479c-8a83-5fabb876a5d6.png">

这里地方是改写了Megatron的数据处理逻辑，可以参照这里https://github.com/NVIDIA/Megatron-LM 的Data Preprocessing的代码生成一份处理好的bin、idx、npy文件。

另外我们有一套自己的数据处理逻辑近期代码会提交，待提交后我再回复您。

**Comment by GGGGGGXY** [comment]
> 请问一下二郎神模型是用fp16训练的吗，我用fp16 finetuing 预训练二郎神模型，loss scale都会不正常， 导致没法训练

在fp16下训练的，可以尝试调整下超参数， 可能是weight_decay这个参数设置不当导致

**Comment by bigprince97** [comment]
谢谢，请问下二郎神预训练时初始化参数的策略是均值0，标准差0.2的正太分布吗，在这种初始化策略下训练梯度会为nan，将标准差改为0.02，才能够正常训练。

**Comment by GGGGGGXY** [comment]
> 谢谢，请问下二郎神预训练时初始化参数的策略是均值0，标准差0.2的正太分布吗，在这种初始化策略下训练梯度会为nan，将标准差改为0.02，才能够正常训练。

是的，用的是0.02初始化
